cloudimg/aws.py

```diff
@@ -44,6 +45,29 @@ class AWSBootMode(enum.Enum):
     """Use the default boot mode from AWS."""
 
 
+def log_request_id(func):
+    """
+    The AWS request ID is useful when troubleshooting errors. When a
+    botocore.exceptions.ClientError error is raised, attempt to log the
+    request ID associated with the error.
+
+    https://boto3.amazonaws.com/v1/documentation/api/latest/guide/error-handling.html#catching-exceptions-when-using-a-low-level-client
+    """
+    @functools.wraps(func)
+    def log_request_id_wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except ClientError as exc:
+            if exc.response.get("ResponseMetadata") and \
+                    exc.response['ResponseMetadata'].get('RequestId'):
+                log.error(
```

Comment:
I'm not sure if this is used from multiple threads, but if it is it has the potential to be ambiguous.  Logs can appear like this:

```
An error occurred in AWS (AWS RequestId: ABC)  # logged from here in thread 1
An error occurred in AWS (AWS RequestId: BCD)  # logged from here in thread 2
<exception details 1>  # logged by something in thread 1 after the 'raise'
<exception details 2>  # logged by something in thread 2 after the 'raise'
```

There would be no way to unambiguously match up the request ID with the exception (unless the exception details already included the request ID, but that would make the change redundant anyway).

This might be better off using `log.exception` to force a single log event which includes both the exception info and the request ID, so there is no ambiguity.

Reply:
Good point, thanks!